UnitySDK/Assets/ML-Agents/Examples/SharedAssets/Scripts/RayPerception3D.cs

```diff
@@ -31,66 +30,26 @@ public class RayPerception3D : RayPerception
         /// <param name="detectableObjects">List of tags which correspond to object types agent can see</param>
         /// <param name="startOffset">Starting height offset of ray from center of agent.</param>
         /// <param name="endOffset">Ending height offset of ray from center of agent.</param>
-        public override List<float> Perceive(float rayDistance,
+        public override IList<float> Perceive(float rayDistance,
             float[] rayAngles, string[] detectableObjects,
             float startOffset=0.0f, float endOffset=0.0f)
         {
-            if (m_SubList == null || m_SubList.Length != detectableObjects.Length + 2)
-                m_SubList = new float[detectableObjects.Length + 2];
-
-            m_PerceptionBuffer.Clear();
-            m_PerceptionBuffer.Capacity = m_SubList.Length * rayAngles.Length;
-
-            // For each ray sublist stores categorical information on detected object
-            // along with object distance.
-            foreach (var angle in rayAngles)
+            var perceptionSize = (detectableObjects.Length + 2) * rayAngles.Length;
+            if (m_PerceptionBuffer == null || m_PerceptionBuffer.Length != perceptionSize)
             {
-                Vector3 startPositionLocal = new Vector3(0, startOffset, 0);
-                Vector3 endPositionLocal = PolarToCartesian(rayDistance, angle);
-                endPositionLocal.y += endOffset;
-
-                var startPositionWorld = transform.TransformPoint(startPositionLocal);
-                var endPositionWorld = transform.TransformPoint(endPositionLocal);
-
-                var rayDirection = endPositionWorld - startPositionWorld;
-                if (Application.isEditor)
-                {
-                    Debug.DrawRay(startPositionWorld,rayDirection, Color.black, 0.01f, true);
-                }
-
-                Array.Clear(m_SubList, 0, m_SubList.Length);
-
-                if (Physics.SphereCast(startPositionWorld, 0.5f, rayDirection, out m_Hit, rayDistance))
-                {
-                    for (var i = 0; i < detectableObjects.Length; i++)
-                    {
-                        if (m_Hit.collider.gameObject.CompareTag(detectableObjects[i]))
-                        {
-                            m_SubList[i] = 1;
-                            m_SubList[detectableObjects.Length + 1] = m_Hit.distance / rayDistance;
-                            break;
-                        }
-                    }
-                }
-                else
-                {
-                    m_SubList[detectableObjects.Length] = 1f;
-                }
-
-                Utilities.AddRangeNoAlloc(m_PerceptionBuffer, m_SubList);
+                m_PerceptionBuffer = new float[perceptionSize];
             }
 
+            const float castRadius = 0.5f;
+            const bool legacyHitFractionBehavior = true;
```

Comment:
I feel this adds complexity. It is fair to say, you need to retrain when updating. I don't think we should have legacy code in a beta project.

Reply:
Agreed it adds complexity. Disagree that we shouldn't support it, at least for one more release. The old code is marked as `[Obsolete]` and we can remove it after the next release ships.